parquet-column/src/main/java/org/apache/parquet/column/impl/ColumnValueCollector.java

```diff
@@ -0,0 +1,139 @@
+import java.util.OptionalLong;
+import org.apache.parquet.column.ColumnDescriptor;
+import org.apache.parquet.column.ParquetProperties;
+import org.apache.parquet.column.statistics.SizeStatistics;
+import org.apache.parquet.column.statistics.Statistics;
+import org.apache.parquet.column.values.bloomfilter.AdaptiveBlockSplitBloomFilter;
+import org.apache.parquet.column.values.bloomfilter.BlockSplitBloomFilter;
+import org.apache.parquet.column.values.bloomfilter.BloomFilter;
+import org.apache.parquet.column.values.bloomfilter.BloomFilterWriter;
+import org.apache.parquet.io.api.Binary;
+
+// An internal class to collect column values to build column statistics and bloom filter.
+class ColumnValueCollector {
+
+  private final ColumnDescriptor path;
+  private final BloomFilterWriter bloomFilterWriter;
+  private final BloomFilter bloomFilter;
+  private Statistics<?> statistics;
+  private SizeStatistics.Builder sizeStatisticsBuilder;
+
+  ColumnValueCollector(ColumnDescriptor path, BloomFilterWriter bloomFilterWriter, ParquetProperties props) {
+    this.path = path;
+    resetPageStatistics();
+
+    this.bloomFilterWriter = bloomFilterWriter;
+    if (bloomFilterWriter == null) {
+      this.bloomFilter = null;
+      return;
+    }
+    int maxBloomFilterSize = props.getMaxBloomFilterBytes();
+
+    OptionalLong ndv = props.getBloomFilterNDV(path);
+    OptionalDouble fpp = props.getBloomFilterFPP(path);
+    // If user specify the column NDV, we construct Bloom filter from it.
+    if (ndv.isPresent()) {
+      int optimalNumOfBits = BlockSplitBloomFilter.optimalNumOfBits(ndv.getAsLong(), fpp.getAsDouble());
+      this.bloomFilter = new BlockSplitBloomFilter(optimalNumOfBits / 8, maxBloomFilterSize);
+    } else {
+      if (props.getAdaptiveBloomFilterEnabled(path)) {
+        int numCandidates = props.getBloomFilterCandidatesCount(path);
+        this.bloomFilter =
+            new AdaptiveBlockSplitBloomFilter(maxBloomFilterSize, numCandidates, fpp.getAsDouble(), path);
+      } else {
+        this.bloomFilter = new BlockSplitBloomFilter(maxBloomFilterSize, maxBloomFilterSize);
+      }
+    }
+  }
+
+  void resetPageStatistics() {
+    this.statistics = Statistics.createStats(path.getPrimitiveType());
+    this.sizeStatisticsBuilder = SizeStatistics.newBuilder(
+        path.getPrimitiveType(), path.getMaxRepetitionLevel(), path.getMaxDefinitionLevel());
+  }
+
+  void writeNull(int repetitionLevel, int definitionLevel) {
+    statistics.incrementNumNulls();
+    sizeStatisticsBuilder.add(repetitionLevel, definitionLevel);
+  }
+
+  void write(boolean value, int repetitionLevel, int definitionLevel) {
+    statistics.updateStats(value);
+    sizeStatisticsBuilder.add(repetitionLevel, definitionLevel);
+  }
+
+  void write(int value, int repetitionLevel, int definitionLevel) {
+    statistics.updateStats(value);
+    sizeStatisticsBuilder.add(repetitionLevel, definitionLevel);
+    if (bloomFilter != null) {
```

Comment:
What do you think about using a no-op BloomFilter implementation instead of a null-check? I am not sure if would perform better, though.